include/modem/lte_lc.h

```diff
@@ -1226,11 +1250,17 @@ struct lte_lc_evt {
 		 */
 		uint64_t time;
 
+#if defined(CONFIG_LTE_LC_NCELLMEAS) || defined(CONFIG_ZTEST) || defined(CONFIG_UNITY) || \
+	defined(__DOXYGEN__)
 		/** Payload for event @ref LTE_LC_EVT_NEIGHBOR_CELL_MEAS. */
 		struct lte_lc_cells_info cells_info;
+#endif
 
+#if defined(CONFIG_LTE_LC_RAI) || defined(CONFIG_ZTEST) || defined(CONFIG_UNITY) || \
+	defined(__DOXYGEN__)
 		/** Payload for event @ref LTE_LC_EVT_RAI_UPDATE. */
 		struct lte_lc_rai_cfg rai_cfg;
+#endif
```

Comment:
Adding a comment here for the functions to get a conversation thread although the line is wrong.
We discussed yesterday about the functions and you said we don't need to flag them as compilation will fail later.

E.g., using `lte_lc_neighbor_cell_measurement()` will fail linking when functions from `ncellmeas.c` are not compiled if `CONFIG_LTE_LC_NCELLMEAS` is not set.

So is this enough of information for developer as the compilation failure will not tell how to fix it. You need to find missing function and see that it's coming from `ncellmeas.c` and then go to `nrf/lib/lte_link_control/modules/CMakeLists.txt` to see that you need to set `CONFIG_LTE_LC_NCELLMEAS` in order to fix linking problem.
This sounds a bit too much. Should we document in the functions what Kconfig option has to be enabled so that they are available? Or flag the functions, or both? This is a provocative question but I feel like we need to do something.

Reply:
Please do not flag headers with CONFIG_* variables. it makes module-testing harder.

Sometimes it is just easier to stub certain modules by turning the Kconfig off and providing a stub in your build. Using the original header makes it obvious if there is any changes in API.

Second:
Use Doxygen grouping or something else for functions that belong to certain modules. It might be too much if each of the fuction has `* Is provided by CONFIG_LTE_LC_THIS_MODULE`.
Instead
```
/**
 * @defgroup lte_lc_this_module LTE Link Controller module for this and that
 * @ingroup lte_lc_this_module
 * @{
 * @brief A this and that module, enabled by CONFIG_LTE_LC_THIS_MODULE
 */

all functions here

/** @} */
```

Reply:
> Please do not flag headers with CONFIG_* variables. it makes module-testing harder.

Does this apply to both functions that I commented, or also to the event type values that are currently flagged in the PR?

Reply:
All.
In my opinion, having `#ifdef` in header is almost always a bad idea.
It should be avoided, unless there is a clear flash saving benefit.

Reply:
I'm not keen on it either. The motivation for this flagging was to avoid clients using an event value in their code and then not having the corresponding module enabled which would mean that the event can never come.
We want compilation errors for cases when the library is misused unintentionally.

Reply:
Good point, but there is a downside.

If certain events are flagged on CONFIG_* values, then it means that definition of `enum lte_lc_evt_type` depends on those configuration macros. And then it means that ALL modules that use `enum lte_lc_evt_type` will also depends on the state of those configuration macros.
The outcome of those circular dependencies is that changing value of any CONFIG_* flag that affects the enumeration values, causes all modules to rebuild. And not only modules, but all libraries that use that `enum lte_lc_evt_type` .

So it also ruins the effect of Ccache as well.

Reply:
I agree with Seppo that conditionally defined header stuff is best avoided -- it doesn't just hurt test writing. Any code that needs to be able to operate with or without the feature now also has to have preprocessor macros littered all around it to make sure it avoids accidentally referencing a potentially undefined type.

What I do sometimes do is replace the function declaration with a dummy macro, or a dummy implementation, when it is not enabled. That way downstream code doesn't need the aggressive macros. This is a pretty common strategy over on the Zephyr side of things, it works quite well IMO. Usually it is best with functions that return an integer code, so that you can return `-ENOTSUPP` or something similar to give a hint to the developer of what went wrong. Or you can even make the dummy implementation print a message indicating the config that needs to be enabled to use the function, but that is less common.

I haven't reviewed this PR yet (waiting for it to reach maturity since I am really not an expert here). These are just my broad thoughts on conditional declarations

Reply:
> I agree with Seppo that conditionally defined header stuff is best avoided -- it doesn't just hurt test writing. Any code that needs to be able to operate with or without the feature now also has to have preprocessor macros littered all around it to make sure it avoids accidentally referencing a potentially undefined type.

As I said, I don't like it either but how do we avoid people misusing the currently flagged enum values? Yes, it is a solution not to flag them and document the needed Kconfig option. I'd still like to see a compilation failure. Especially, given that we are making a backwards incompatible change to disable functionality that was earlier always enabled.

> What I sometimes do is replace the function declaration with a dummy macro, or a dummy implementation, when it is not enabled.
> That way downstream code doesn't need the aggressive macros. This is a pretty common strategy over on the Zephyr side of things, it works quite well IMO. Usually it is best with functions that return an integer code, so that you can return `-ENOTSUPP` or something similar to give a hint to the developer of what went wrong. Or you can even make the dummy implementation print a message indicating the config that needs to be enabled to use the function, but that is less common.

The downside is that it doesn't give compiler error. With current implementation, we at lest get compilation failure although the error message doesn't really tell how to fix it but you need to browse the code to figure out which Kconfig to enable. That can be improved by documenting them.
Asserting and printing how to fix the issue is possible and helps a bit.

I would expect that clients don't need flagging other than `CONFIG_LTE_LINK_CONTROL` if they have one because they mostly use certain features and can enable necessary flags in the `prj.conf` or related overlay files.

> I haven't reviewed this PR yet (waiting for it to reach maturity since I am really not an expert here). These are just my broad thoughts on conditional declarations

Please review it ASAP to get this into NCS 2.8.0.

Reply:
> All. In my opinion, having `#ifdef` in header is almost always a bad idea. It should be avoided, unless there is a clear flash saving benefit.

I just want to clarify that this PR is indeed for reducing flash usage, the modular architecture is a byproduct of that. By default, right now, any application enabling the LTE Link Controller is also building the full set of AT monitors, even when the application is simply doing a connect.

Reply:
> The outcome of those circular dependencies is that changing value of any CONFIG_* flag that affects the enumeration values, causes all modules to rebuild. And not only modules, but all libraries that use that enum lte_lc_evt_type .

@SeppoTakalo I don't think this is a big problem, after all changing the LTE LC module flagging and rebuilding the code isn't something that is done regularly. And our applications are so small, that a full rebuild doesn't take that long anyway.

I don't particularly like flagging in headers either, but I think there has to be a compile time error if the application is expecting an event which due to a configuration error will never come. Especially since we're changing the default configuration.

Looking at Zephyr headers, Kconfig flagging seems to be widely used there.

Reply:
We are currently going ahead by flagging the event enum values and the respective data structure with the absence of a better solution to provide compilation failure. We are not flagging the functions though.
I will close this topic after the silence. If you have further comments, please open and add further comments into this conversation.